File: src/io/jbeadmemento.cpp
#include "jbeadmemento.h"

#include "jbeadstream.h"
#include "objectmodel.h"
#include "domain/beadsymbols.h"

namespace jbead {

void JBeadMemento::save(QIODevice* out) const
{
    /*  File layout (S-expression):

          (jbb
            (version 1)
            (author "...")
            (organization "...")
            (notes "...")
            (colors
              (rgb r g b a)
              ... one per palette entry ...)
            (view
              (draft-visible bool) (corrected-visible bool)
              (simulation-visible bool) (report-visible bool)
              (selected-tool "pencil"|"select"|...)
              (selected-color int) (zoom int)
              (scroll int) (shift int)
              (draw-colors bool) (draw-symbols bool)
              (symbols "..."))
            (model
              (row b0 b1 ... b_{width-1})
              ... height rows total ...))                          */

    ObjectModel om(QStringLiteral("jbb"));

    om.add(QStringLiteral("version"),      {VERSION});
    om.add(QStringLiteral("author"),       {m_author});
    om.add(QStringLiteral("organization"), {m_organization});
    om.add(QStringLiteral("notes"),        {m_notes});

    for (const QColor& c : m_colors) {
        om.add(QStringLiteral("colors/rgb"),
               {c.red(), c.green(), c.blue(), c.alpha()});
    }

    om.add(QStringLiteral("view/draft-visible"),      {m_draftVisible});
    om.add(QStringLiteral("view/corrected-visible"),  {m_correctedVisible});
    om.add(QStringLiteral("view/simulation-visible"), {m_simulationVisible});
    om.add(QStringLiteral("view/report-visible"),     {m_reportVisible});
    om.add(QStringLiteral("view/selected-tool"),      {m_selectedTool});
    om.add(QStringLiteral("view/selected-color"),     {int(m_colorIndex)});
    om.add(QStringLiteral("view/zoom"),               {m_zoomIndex});
    om.add(QStringLiteral("view/scroll"),             {m_scroll});
    om.add(QStringLiteral("view/shift"),              {m_shift});
    om.add(QStringLiteral("view/draw-colors"),        {m_drawColors});
    om.add(QStringLiteral("view/draw-symbols"),       {m_drawSymbols});
    om.add(QStringLiteral("view/symbols"),            {m_symbols});

    for (int j = 0; j < m_height; ++j) {
        QList<QVariant> row;
        row.reserve(m_width);
        for (int i = 0; i < m_width; ++i) {
            row.append(int(static_cast<std::int8_t>(m_data.at(j * m_width + i))));
        }
        om.add(QStringLiteral("model/row"), row);
    }

    JBeadOutputStream s(out);
    s.write(om.toString());
}

void JBeadMemento::load(QIODevice* in)
{
    JBeadInputStream s(in);
    const ObjectModel om = ObjectModel::fromData(s.readAll(), QStringLiteral("jbb"));

    /*  version is currently always 1; the legacy upgrade() path is
        a no-op stub. Honour the same intent: read the field and
        ignore the version number for now.                         */
    (void) om.intValue(QStringLiteral("version"), 1);
    m_author       = om.stringValue(QStringLiteral("author"), QString());
    m_organization = om.stringValue(QStringLiteral("organization"), QString());
    m_notes        = om.stringValue(QStringLiteral("notes"), QString());

    m_colors.clear();
    for (Node* n : om.getAll(QStringLiteral("colors/rgb"))) {
        if (!n->isLeaf()) continue;
        Leaf* leaf = static_cast<Leaf*>(n);
        const int r = leaf->intValue(0);
        const int g = leaf->intValue(1);
        const int b = leaf->intValue(2);
        const int a = leaf->valueCount() == 4 ? leaf->intValue(3) : 255;
        m_colors.append(QColor(r, g, b, a));
    }

    m_draftVisible      = om.boolValue(QStringLiteral("view/draft-visible"),      true);
    m_correctedVisible  = om.boolValue(QStringLiteral("view/corrected-visible"),  true);
    m_simulationVisible = om.boolValue(QStringLiteral("view/simulation-visible"), true);
    m_reportVisible     = om.boolValue(QStringLiteral("view/report-visible"),     true);
    m_colorIndex   = static_cast<std::int8_t>(om.intValue(QStringLiteral("view/selected-color"), 1));
    m_selectedTool = om.stringValue(QStringLiteral("view/selected-tool"), QStringLiteral("pencil"));
    m_zoomIndex    = om.intValue(QStringLiteral("view/zoom"),   2);
    m_scroll       = om.intValue(QStringLiteral("view/scroll"), 0);
    m_shift        = om.intValue(QStringLiteral("view/shift"),  0);
    m_drawColors   = om.boolValue(QStringLiteral("view/draw-colors"),  true);
    m_drawSymbols  = om.boolValue(QStringLiteral("view/draw-symbols"), false);
    m_symbols      = om.stringValue(QStringLiteral("view/symbols"), BeadSymbols::DEFAULT_SYMBOLS);

    const QList<Node*> rows = om.getAll(QStringLiteral("model/row"));
    m_height = rows.size();
    m_width  = rows.isEmpty() ? 0 : static_cast<Leaf*>(rows.first())->valueCount();
    m_data.resize(m_width * m_height);
    int idx = 0;
    for (Node* row : rows) {
        Leaf* leaf = static_cast<Leaf*>(row);
        for (int i = 0; i < m_width; ++i) {
            m_data[idx++] = static_cast<char>(leaf->intValue(i) & 0xff);
        }
    }
}

} // namespace jbead

File: legacy/src/ch/jbead/fileformat/JBeadMemento.java
/** jbead - http://www.jbead.ch
    Copyright (C) 2001-2012  Damian Brunold

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package ch.jbead.fileformat;

import java.awt.Color;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import ch.jbead.BeadSymbols;
import ch.jbead.storage.Node;
import ch.jbead.storage.ObjectModel;

public class JBeadMemento extends Memento {

    public static final int VERSION = 1;

    @Override
    public void save(JBeadOutputStream out) throws IOException {
        ObjectModel om = new ObjectModel("jbb");
        saveInfos(om);
        saveColors(om);
        saveView(om);
        savePattern(om);
        out.write(om.toString());
    }

    private void saveInfos(ObjectModel om) {
        om.add("version", VERSION);
        om.add("author", author);
        om.add("organization", organization);
        om.add("notes", notes);
    }

    private void saveColors(ObjectModel om) {
        for (Color color: colors) {
            om.add("colors/rgb", color.getRed(), color.getGreen(), color.getBlue(), color.getAlpha());
        }
    }

    private void saveView(ObjectModel om) {
        om.add("view/draft-visible", draftVisible);
        om.add("view/corrected-visible", correctedVisible);
        om.add("view/simulation-visible", simulationVisible);
        om.add("view/report-visible", reportVisible);
        om.add("view/selected-tool", selectedTool);
        om.add("view/selected-color", colorIndex);
        om.add("view/zoom", zoomIndex);
        om.add("view/scroll", scroll);
        om.add("view/shift", shift);
        om.add("view/draw-colors", drawColors);
        om.add("view/draw-symbols", drawSymbols);
        om.add("view/symbols", symbols);
    }

    private void savePattern(ObjectModel om) {
        for (int j = 0; j < height; j++) {
            List<Byte> row = new ArrayList<Byte>();
            for (int i = 0; i < width; i++) {
                row.add(data[j * width + i]);
            }
            om.add("model/row", row.toArray());
        }
    }

    @Override
    public void load(JBeadInputStream in) throws IOException {
        ObjectModel om = ObjectModel.fromData(in.readAll());
        loadInfos(om);
        loadColors(om);
        loadView(om);
        loadPattern(om);
    }

    private void loadInfos(ObjectModel om) {
        int version = om.getIntValue("version", 1);
        if (version < VERSION) {
            upgrade(om, version);
        }
        author = (String) om.getStringValue("author", "");
        organization = (String) om.getStringValue("organization", "");
        notes = (String) om.getStringValue("notes", "");
    }

    private void loadColors(ObjectModel om) {
        colors.clear();
        for (Node color : om.getAll("colors/rgb")) {
            colors.add(getColor(color));
        }
    }

    private void loadView(ObjectModel om) {
        draftVisible = om.getBoolValue("view/draft-visible", true);
        correctedVisible = om.getBoolValue("view/corrected-visible", true);
        simulationVisible = om.getBoolValue("view/simulation-visible", true);
        reportVisible = om.getBoolValue("view/report-visible", true);
        colorIndex = (byte) om.getIntValue("view/selected-color", 1);
        selectedTool = om.getStringValue("view/selected-tool", "pencil");
        zoomIndex = om.getIntValue("view/zoom", 2);
        scroll = om.getIntValue("view/scroll", 0);
        shift = om.getIntValue("view/shift", 0);
        drawColors = om.getBoolValue("view/draw-colors", true);
        drawSymbols = om.getBoolValue("view/draw-symbols", false);
        symbols = om.getStringValue("view/symbols", BeadSymbols.SAVED_SYMBOLS);
    }

    private void loadPattern(ObjectModel om) {
        List<Node> rows = om.getAll("model/row");
        height = rows.size();
        width = rows.get(0).size();
        data = new byte[width * height];
        int idx = 0;
        for (Node row : rows) {
            for (int i = 0; i < width; i++) {
                data[idx++] = (byte) row.asLeaf().getIntValue(i);
            }
        }
    }
    private void upgrade(ObjectModel om, int version) {
        // in the future, here will be conversion code that upgrades from earlier versions
    }

    private Color getColor(Node color) {
        int red = color.asLeaf().getIntValue(0);
        int green = color.asLeaf().getIntValue(1);
        int blue = color.asLeaf().getIntValue(2);
        if (color.asLeaf().size() == 4) {
            int alpha = color.asLeaf().getIntValue(3);
            return new Color(red, green, blue, alpha);
        } else {
            return new Color(red, green, blue);
        }
    }

}
